drivers/i3c/i3c_npcx.c

```diff
@@ -2057,63 +2464,405 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
+	uint8_t bamatch;
+	int idx_module = GET_MODULE_ID(config->instance_id);
+	int ret;
+	uint64_t pid;
+
+	/* I3C module mdma cotroller or target mode select */
+	npcx_i3c_target_sel(idx_module, true);
+
+	/* Set bus available match value in target register */
+	ret = clock_control_get_rate(clk_dev, (clock_control_subsys_t)&config->ref_clk_subsys,
+				     &apb4_rate);
+	LOG_DBG("APB4_CLK: %d", apb4_rate);
+
+	if (ret != 0x0) {
+		LOG_ERR("Get APB4 source clock fail %d", ret);
+		return -EINVAL;
+	}
+
+	bamatch = get_bus_available_match_val(apb4_rate);
+	LOG_DBG("BAMATCH: %d", bamatch);
+	SET_FIELD(inst->CONFIG, NPCX_I3C_CONFIG_BAMATCH, bamatch);
+
+	/* Enable secondary controller capability if supported */
+	if (config->secondary_mode == true) {
+		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_CAPABLE);
+	} else {
+		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_OFF);
+	}
+
+	/* Set Provisional ID */
+	pid = config_target->pid;
+
+	/* PID[47:33] MIPI manufacturer ID */
+	SET_FIELD(inst->VENDORID, NPCX_I3C_VENDORID_VID, (uint32_t)GET_PID_VENDOR_ID(pid));
+
+	/* PID[32] Verdor fixed value(0) or random value(1) */
+	if (config_target->pid_random) {
+		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_IDRAND);
+	} else {
+		inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_IDRAND);
+	}
+
+	/* PID[31:0] vendor fixed value */
+	inst->PARTNO = (uint32_t)GET_PID_PARTNO(pid);
+
+	LOG_DBG("pid: %#llx", pid);
+	LOG_DBG("vendro id: %#x", (uint32_t)GET_PID_VENDOR_ID(pid));
+	LOG_DBG("id type: %d", (uint32_t)GET_PID_ID_TYP(pid));
+	LOG_DBG("partno: %#x", (uint32_t)GET_PID_PARTNO(pid));
+
+	SET_FIELD(inst->IDEXT, NPCX_I3C_IDEXT_DCR, config_target->dcr);
+	SET_FIELD(inst->IDEXT, NPCX_I3C_IDEXT_BCR, config_target->bcr);
+	SET_FIELD(inst->CONFIG, NPCX_I3C_CONFIG_SADDR, config_target->static_addr);
+	SET_FIELD(inst->CONFIG, NPCX_I3C_CONFIG_HDRCMD, CFG_HDRCMD_RD_FROM_FIFIO);
+	SET_FIELD(inst->MAXLIMITS, NPCX_I3C_MAXLIMITS_MAXRD, (config_target->max_read_len) & 0xfff);
+	SET_FIELD(inst->MAXLIMITS, NPCX_I3C_MAXLIMITS_MAXWR,
+		  (config_target->max_write_len) & 0xfff);
+
+	/* Ignore DA and detect all START and STOP */
+	inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_MATCHSS);
+	inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Enable target device */
+
+	/* Enable the target interrupt events */
+	npcx_i3c_enable_target_interrupt(dev, true);
+
+	return 0;
+}
+
 static int npcx_i3c_configure(const struct device *dev, enum i3c_config_type type, void *config)
 {
 	struct npcx_i3c_data *dev_data = dev->data;
-	struct i3c_config_controller *cntlr_cfg = config;
+	struct i3c_config_controller *config_cntlr;
+	struct i3c_config_target *config_target;
 
 	if (type == I3C_CONFIG_CONTROLLER) {
+		config_cntlr = config;
 		/*
 		 * Check for valid configuration parameters.
 		 * Currently, must be the primary controller.
 		 */
-		if ((cntlr_cfg->is_secondary) || (cntlr_cfg->scl.i3c == 0U)) {
+		if ((config_cntlr->is_secondary) || (config_cntlr->scl.i3c == 0U)) {
```

Comment:
You should still be able to configure as a controller while a secondary controller.... although Handoff support isn't quite here yet